agent/recorder.go

```diff
@@ -167,19 +181,53 @@ func (r *SpanRecorder) SendSpans() (error, bool) {
 	}
 
 	if payloadSent {
-		r.okSend++
+		atomic.AddInt64(&r.stats.sendSpansOk, 1)
+		atomic.AddInt64(&r.stats.spansSent, int64(len(spans)))
 	} else {
-		r.koSend++
+		atomic.AddInt64(&r.stats.sendSpansKo, 1)
+		atomic.AddInt64(&r.stats.spansNotSent, int64(len(spans)))
 	}
 
 	return lastError, shouldExit
 }
 
+// Stop recorder
+func (r *SpanRecorder) Stop() {
+	if r.debugMode {
+		r.logger.Println("Scope recorder is stopping gracefully...")
+	}
+	r.t.Kill(nil)
+	_ = r.t.Wait()
+	if r.hasSpans() {
+		err, _ := r.sendSpans()
+		if err != nil {
+			r.logger.Printf("error sending spans: %v\n", err)
+		}
+	}
+	if r.debugMode {
+		r.writeStats()
+	}
+}
+
+// Write statistics
+func (r *SpanRecorder) writeStats() {
+	r.statsOnce.Do(func() {
+		r.logger.Printf("** Recorder statistics **\n")
+		r.logger.Printf("  Total spans: %d\n", r.stats.totalSpans)
+		r.logger.Printf("  Test spans: %d\n", r.stats.testSpans)
+		r.logger.Printf("  Spans sent: %d\n", r.stats.spansSent)
+		r.logger.Printf("  Spans not sent: %d\n", r.stats.spansNotSent)
+		r.logger.Printf("  Spans rejected: %d\n", r.stats.spansRejected)
+		r.logger.Printf("  SendSpans calls: %d\n", r.stats.sendSpansCalls)
+		r.logger.Printf("  SendSpans OK: %d\n", r.stats.sendSpansOk)
+		r.logger.Printf("  SendSpans KO: %d\n", r.stats.sendSpansKo)
+	})
+}
+
 // Sends the encoded `payload` to the Scope ingest endpoint
 func (r *SpanRecorder) callIngest(payload io.Reader) (statusCode int, err error) {
```

Comment:
I think all retry logic should be in this function. This should return just `error`, either an unrecoverable error, or the error of the last retry. It will simplify the error handling that at the moment is split between this function and `sendSpans`

Reply:
ok, moving the retry logic here

Reply:
done in the last commit